File: btc_options_fullstack/btc-options/backend/app/services/indicators.py
"""Technical indicator computation for the historical/backtest charts.

Public API: ``compute_indicators(df, configs)`` — takes an OHLCV DataFrame and
a list of indicator configs, returns a dict keyed by stable indicator IDs.

All indicators are hand-rolled in pandas/numpy (no pandas-ta) to keep the
dependency footprint minimal and avoid the known pandas-ta + numpy 2.x
incompatibility (numpy.NaN was removed). Each indicator is ~5 lines.
"""

from __future__ import annotations

import logging
from typing import Any

import numpy as np
import pandas as pd

logger = logging.getLogger(__name__)


# ── Stable indicator IDs ──────────────────────────────────────────────────────

def indicator_id(cfg: dict) -> str:
    """Build a stable string ID like 'rsi_14', 'ema_20', 'macd_12_26_9'."""
    t = cfg.get("type", "")
    p = cfg.get("params", {}) or {}
    if t in ("sma", "ema", "rsi", "atr"):
        return f"{t}_{int(p.get('period', 0))}"
    if t == "bbands":
        return f"bbands_{int(p.get('period', 20))}_{int(p.get('stdev', 2))}"
    if t == "macd":
        return f"macd_{int(p.get('fast', 12))}_{int(p.get('slow', 26))}_{int(p.get('signal', 9))}"
    if t == "vwap":
        return "vwap"
    return f"{t}_{p}"


# ── Per-indicator handlers ────────────────────────────────────────────────────

def _sma(close: pd.Series, period: int) -> pd.Series:
    return close.rolling(window=period, min_periods=period).mean()


def _ema(close: pd.Series, period: int) -> pd.Series:
    # Standard EMA (Wilder/regular). adjust=False for the recursive form.
    return close.ewm(span=period, adjust=False, min_periods=period).mean()


def _rsi(close: pd.Series, period: int) -> pd.Series:
    """Wilder's RSI."""
    delta = close.diff()
    gain = delta.clip(lower=0.0)
    loss = -delta.clip(upper=0.0)
    avg_gain = gain.ewm(alpha=1 / period, adjust=False, min_periods=period).mean()
    avg_loss = loss.ewm(alpha=1 / period, adjust=False, min_periods=period).mean()
    rs = avg_gain / avg_loss.replace(0.0, np.nan)
    rsi = 100.0 - (100.0 / (1.0 + rs))
    return rsi


def _macd(close: pd.Series, fast: int, slow: int, signal: int):
    fast_ema = close.ewm(span=fast, adjust=False, min_periods=fast).mean()
    slow_ema = close.ewm(span=slow, adjust=False, min_periods=slow).mean()
    macd_line = fast_ema - slow_ema
    sig_line = macd_line.ewm(span=signal, adjust=False, min_periods=signal).mean()
    hist = macd_line - sig_line
    return macd_line, sig_line, hist


def _bbands(close: pd.Series, period: int, stdev: float):
    mid = close.rolling(window=period, min_periods=period).mean()
    std = close.rolling(window=period, min_periods=period).std(ddof=0)
    upper = mid + stdev * std
    lower = mid - stdev * std
    return upper, mid, lower


def _atr(high: pd.Series, low: pd.Series, close: pd.Series, period: int) -> pd.Series:
    """Wilder's ATR."""
    prev_close = close.shift(1)
    tr = pd.concat([
        (high - low).abs(),
        (high - prev_close).abs(),
        (low - prev_close).abs(),
    ], axis=1).max(axis=1)
    return tr.ewm(alpha=1 / period, adjust=False, min_periods=period).mean()


def _vwap(close: pd.Series, high: pd.Series, low: pd.Series, volume: pd.Series) -> pd.Series:
    """Cumulative VWAP using typical-price × volume.

    NOTE: classic intraday VWAP resets each session; here we compute over the
    requested window (caller decides). For intraday use, request a single-day
    window. Returns NaN if volume is all-zero/missing.
    """
    if volume is None or volume.fillna(0).sum() == 0:
        return pd.Series(np.nan, index=close.index)
    tp = (high + low + close) / 3.0
    cum_pv = (tp * volume).cumsum()
    cum_v = volume.cumsum()
    return cum_pv / cum_v.replace(0.0, np.nan)


# ── Public entry point ────────────────────────────────────────────────────────

def compute_indicators(
    df: pd.DataFrame,
    configs: list[dict],
) -> dict[str, list[dict[str, Any]]]:
    """Compute requested indicators on ``df``.

    Args:
        df: must have columns ``time`` (unix sec int), ``close``; optionally
            ``open``, ``high``, ``low``, ``volume``. Sorted ascending by time.
        configs: e.g. [{"type": "rsi", "params": {"period": 14}}, ...].

    Returns:
        {indicator_id: [{time, ...values}, ...]} — one list of points per id.
        Single-value indicators yield ``{time, value}``; MACD yields
        ``{time, macd, signal, hist}``; BBands yields ``{time, upper, mid, lower}``.
        NaN bars (warm-up) are dropped from each indicator's series.
    """
    out: dict[str, list[dict]] = {}
    if df.empty or not configs:
        return out

    close = df["close"].astype(float)
    high  = df["high"].astype(float)  if "high"   in df.columns else close
    low   = df["low"].astype(float)   if "low"    in df.columns else close
    vol   = df["volume"].astype(float) if "volume" in df.columns else None
    times = df["time"].astype(int).values

    for cfg in configs:
        t = (cfg.get("type") or "").lower()
        p = cfg.get("params", {}) or {}
        try:
            iid = indicator_id(cfg)
            if t == "sma":
                series = _sma(close, int(p.get("period", 20)))
                out[iid] = _to_value_points(times, series)
            elif t == "ema":
                series = _ema(close, int(p.get("period", 20)))
                out[iid] = _to_value_points(times, series)
            elif t == "rsi":
                series = _rsi(close, int(p.get("period", 14)))
                out[iid] = _to_value_points(times, series)
            elif t == "atr":
                series = _atr(high, low, close, int(p.get("period", 14)))
                out[iid] = _to_value_points(times, series)
            elif t == "vwap":
                series = _vwap(close, high, low, vol) if vol is not None else pd.Series(np.nan, index=close.index)
                out[iid] = _to_value_points(times, series)
            elif t == "macd":
                m, s, h = _macd(close,
                                int(p.get("fast", 12)),
                                int(p.get("slow", 26)),
                                int(p.get("signal", 9)))
                out[iid] = _to_macd_points(times, m, s, h)
            elif t == "bbands":
                u, mid, lo = _bbands(close,
                                     int(p.get("period", 20)),
                                     float(p.get("stdev", 2)))
                out[iid] = _to_bbands_points(times, u, mid, lo)
            else:
                logger.warning(f"Unknown indicator type: {t}")
        except Exception as e:
            logger.warning(f"Indicator {cfg} failed: {e}")
    return out


# ── Helpers: Series → list of points (drop NaN warm-up bars) ──────────────────

def _to_value_points(times, s: pd.Series) -> list[dict]:
    arr = s.values
    return [
        {"time": int(times[i]), "value": float(arr[i])}
        for i in range(len(arr))
        if not np.isnan(arr[i])
    ]


def _to_macd_points(times, m: pd.Series, s: pd.Series, h: pd.Series) -> list[dict]:
    ma, sa, ha = m.values, s.values, h.values
    out = []
    for i in range(len(ma)):
        if np.isnan(ma[i]) and np.isnan(sa[i]) and np.isnan(ha[i]):
            continue
        out.append({
            "time": int(times[i]),
            "macd":   None if np.isnan(ma[i]) else float(ma[i]),
            "signal": None if np.isnan(sa[i]) else float(sa[i]),
            "hist":   None if np.isnan(ha[i]) else float(ha[i]),
        })
    return out


def _to_bbands_points(times, u: pd.Series, m: pd.Series, lo: pd.Series) -> list[dict]:
    ua, ma, la = u.values, m.values, lo.values
    out = []
    for i in range(len(ua)):
        if np.isnan(ua[i]) and np.isnan(ma[i]) and np.isnan(la[i]):
            continue
        out.append({
            "time":  int(times[i]),
            "upper": None if np.isnan(ua[i]) else float(ua[i]),
            "mid":   None if np.isnan(ma[i]) else float(ma[i]),
            "lower": None if np.isnan(la[i]) else float(la[i]),
        })
    return out

File: btc_options_fullstack/btc-options/frontend/src/utils/indicators.ts
// TS port of `backend/app/services/indicators.py` for client-side indicator
// computation on M7TradePathChart spot OHLC.
//
// Parity with backend:
//   - `_sma`, `_ema`, `_rsi` (Wilder), `_macd`, `_bbands`, `_atr` (Wilder)
//   - Same warm-up (min_periods semantics → NaN for first N-1 bars)
//   - EMA uses the recursive form with `alpha = 2 / (period + 1)`,
//     seeded with the first available close (matches pandas
//     `ewm(span=N, adjust=False, min_periods=N).mean()` from period N onward).
//   - RSI uses Wilder's smoothing: `alpha = 1 / period` on gains/losses.
//   - ATR uses Wilder's smoothing on True Range.
//
// The parity test in `indicators.test.ts` confirms RSI/MACD/BB/ATR match the
// backend within 1e-6 on a fixed 60-bar synthetic series.

export interface OhlcBar {
  time: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number;
}

export interface ValuePoint { time: number; value: number; }
export interface MacdPoint { time: number; macd: number; signal: number; hist: number; }
export interface BBandsPoint { time: number; upper: number; mid: number; lower: number; }

// ── Primitives ──────────────────────────────────────────────────────────────

/** Simple moving average — first `period - 1` bars are NaN. */
export function sma(close: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(close.length).fill(null);
  if (close.length < period) return out;
  let sum = 0;
  for (let i = 0; i < period; i++) sum += close[i];
  out[period - 1] = sum / period;
  for (let i = period; i < close.length; i++) {
    sum += close[i] - close[i - period];
    out[i] = sum / period;
  }
  return out;
}

/** Exponential moving average — recursive form. Seeded at index period-1
 *  with SMA over the first `period` values, NaN elsewhere. Matches pandas
 *  `ewm(span=N, adjust=False, min_periods=N).mean()`. */
export function ema(close: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(close.length).fill(null);
  if (close.length < period) return out;
  const alpha = 2 / (period + 1);
  // Seed with SMA of first N values
  let seed = 0;
  for (let i = 0; i < period; i++) seed += close[i];
  out[period - 1] = seed / period;
  for (let i = period; i < close.length; i++) {
    out[i] = alpha * close[i] + (1 - alpha) * (out[i - 1] as number);
  }
  return out;
}

/** Wilder's RSI. Output is in [0, 100]; NaN until period bars elapsed. */
export function rsi(close: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(close.length).fill(null);
  if (close.length <= period) return out;
  const alpha = 1 / period;
  let avgGain = 0, avgLoss = 0;
  // Wilder seed: average gain/loss over first `period` deltas
  for (let i = 1; i <= period; i++) {
    const d = close[i] - close[i - 1];
    avgGain += d > 0 ? d : 0;
    avgLoss += d < 0 ? -d : 0;
  }
  avgGain /= period;
  avgLoss /= period;
  out[period] = avgLoss === 0 ? 100 : 100 - (100 / (1 + avgGain / avgLoss));
  for (let i = period + 1; i < close.length; i++) {
    const d = close[i] - close[i - 1];
    const gain = d > 0 ? d : 0;
    const loss = d < 0 ? -d : 0;
    avgGain = alpha * gain + (1 - alpha) * avgGain;
    avgLoss = alpha * loss + (1 - alpha) * avgLoss;
    out[i] = avgLoss === 0 ? 100 : 100 - (100 / (1 + avgGain / avgLoss));
  }
  return out;
}

/** MACD = EMA(fast) − EMA(slow); signal = EMA(MACD, signal); hist = MACD − signal. */
export function macd(
  close: number[],
  fast = 12, slow = 26, signal = 9,
): { macd: (number | null)[]; signal: (number | null)[]; hist: (number | null)[]; } {
  const fastEma = ema(close, fast);
  const slowEma = ema(close, slow);
  const macdLine: (number | null)[] = close.map((_, i) =>
    (fastEma[i] != null && slowEma[i] != null)
      ? (fastEma[i] as number) - (slowEma[i] as number) : null);

  // Signal = EMA over the macd line, but we have to handle the head NaNs.
  const sigLine: (number | null)[] = new Array(close.length).fill(null);
  const macdNumeric = macdLine.map(v => v as number | null);
  let firstValidIdx = -1;
  for (let i = 0; i < macdNumeric.length; i++) {
    if (macdNumeric[i] != null) { firstValidIdx = i; break; }
  }
  if (firstValidIdx >= 0 && (macdNumeric.length - firstValidIdx) >= signal) {
    const alpha = 2 / (signal + 1);
    let seed = 0;
    for (let i = 0; i < signal; i++) seed += macdNumeric[firstValidIdx + i] as number;
    sigLine[firstValidIdx + signal - 1] = seed / signal;
    for (let i = firstValidIdx + signal; i < close.length; i++) {
      sigLine[i] = alpha * (macdNumeric[i] as number) +
                   (1 - alpha) * (sigLine[i - 1] as number);
    }
  }

  const hist: (number | null)[] = macdLine.map((v, i) =>
    (v != null && sigLine[i] != null)
      ? (v as number) - (sigLine[i] as number) : null);
  return { macd: macdLine, signal: sigLine, hist };
}

/** Bollinger Bands: mid = SMA, upper = mid + k·σ, lower = mid − k·σ.
 *  σ uses ddof=0 (population std). */
export function bbands(
  close: number[], period = 20, stdev = 2,
): { upper: (number | null)[]; mid: (number | null)[]; lower: (number | null)[]; } {
  const mid = sma(close, period);
  const upper: (number | null)[] = new Array(close.length).fill(null);
  const lower: (number | null)[] = new Array(close.length).fill(null);
  for (let i = period - 1; i < close.length; i++) {
    if (mid[i] == null) continue;
    let sumSq = 0;
    const m = mid[i] as number;
    for (let j = i - period + 1; j <= i; j++) sumSq += (close[j] - m) ** 2;
    const sigma = Math.sqrt(sumSq / period);
    upper[i] = m + stdev * sigma;
    lower[i] = m - stdev * sigma;
  }
  return { upper, mid, lower };
}

/** Wilder's ATR (Average True Range). */
export function atr(
  high: number[], low: number[], close: number[], period = 14,
): (number | null)[] {
  const out: (number | null)[] = new Array(close.length).fill(null);
  if (close.length <= period) return out;
  const tr: number[] = [0]; // First bar has no prior close → TR=high-low only
  tr[0] = high[0] - low[0];
  for (let i = 1; i < close.length; i++) {
    const a = high[i] - low[i];
    const b = Math.abs(high[i] - close[i - 1]);
    const c = Math.abs(low[i] - close[i - 1]);
    tr.push(Math.max(a, b, c));
  }
  // Wilder smoothing seeded with average of first `period` TRs.
  let avg = 0;
  for (let i = 0; i < period; i++) avg += tr[i];
  avg /= period;
  out[period - 1] = avg;
  const alpha = 1 / period;
  for (let i = period; i < close.length; i++) {
    avg = alpha * tr[i] + (1 - alpha) * avg;
    out[i] = avg;
  }
  return out;
}

// ── Convenience wrappers that emit time-aligned points ──────────────────────

export function smaSeries(bars: OhlcBar[], period: number): ValuePoint[] {
  return _toValuePoints(bars, sma(bars.map(b => b.close), period));
}
export function emaSeries(bars: OhlcBar[], period: number): ValuePoint[] {
  return _toValuePoints(bars, ema(bars.map(b => b.close), period));
}
export function rsiSeries(bars: OhlcBar[], period: number): ValuePoint[] {
  return _toValuePoints(bars, rsi(bars.map(b => b.close), period));
}
export function atrSeries(bars: OhlcBar[], period: number): ValuePoint[] {
  return _toValuePoints(bars,
    atr(bars.map(b => b.high), bars.map(b => b.low), bars.map(b => b.close), period));
}
export function macdSeries(bars: OhlcBar[], fast = 12, slow = 26, signal = 9): MacdPoint[] {
  const close = bars.map(b => b.close);
  const r = macd(close, fast, slow, signal);
  const out: MacdPoint[] = [];
  for (let i = 0; i < bars.length; i++) {
    if (r.macd[i] == null || r.signal[i] == null || r.hist[i] == null) continue;
    out.push({ time: bars[i].time,
               macd:   r.macd[i]   as number,
               signal: r.signal[i] as number,
               hist:   r.hist[i]   as number });
  }
  return out;
}
export function bbandsSeries(bars: OhlcBar[], period = 20, stdev = 2): BBandsPoint[] {
  const r = bbands(bars.map(b => b.close), period, stdev);
  const out: BBandsPoint[] = [];
  for (let i = 0; i < bars.length; i++) {
    if (r.upper[i] == null || r.mid[i] == null || r.lower[i] == null) continue;
    out.push({ time: bars[i].time,
               upper: r.upper[i] as number,
               mid:   r.mid[i]   as number,
               lower: r.lower[i] as number });
  }
  return out;
}

// ── Internal ─────────────────────────────────────────────────────────────────

function _toValuePoints(bars: OhlcBar[], series: (number | null)[]): ValuePoint[] {
  const out: ValuePoint[] = [];
  for (let i = 0; i < bars.length; i++) {
    const v = series[i];
    if (v == null || !Number.isFinite(v)) continue;
    out.push({ time: bars[i].time, value: v });
  }
  return out;
}
